### PR TITLE
Add GridGenerator for a general quadrilateral/general hexahedron.

### PR DESCRIPTION
**Pull request by Rombur**
I am not happy with the name of the function, so if you have a better one let me know.

### CODE REVIEW

**Comment by bangerth** [comment]
Maybe single_cell?

File: include/deal.II/grid/grid_generator.h [hunk 1]
```diff
@@ -292,6 +292,20 @@ namespace GridGenerator
           const std::vector<unsigned int> &holes);
 
   /**
+   * A general quadrilateral in 2d or a general hexahedron in 3d. It is the
+   * responsability of the user to provide the vertices in the right order (see
```
Comment (tjhei) [inline include/deal.II/grid/grid_generator.h]: responsibility

File: include/deal.II/grid/grid_generator.h [hunk 2]
```diff
@@ -292,6 +292,20 @@ namespace GridGenerator
...
+   * the documentation of the GeometryInfo class).
+   *
```
Comment (tjhei) [inline include/deal.II/grid/grid_generator.h]: document ``colorize``

File: include/deal.II/grid/grid_generator.h [hunk 3]
```diff
@@ -292,6 +292,20 @@ namespace GridGenerator
...
+   * @author Bruno Turcksin
+   * @date 2016
+   */
+  template <int dim>
+  void
+  hyper_quadrilateral(Triangulation<dim> &tria,
```
Comment (tjhei) [inline include/deal.II/grid/grid_generator.h]: I don't like the name, some ideas: ``single_cell``, ``general_cell``, ``quad_or_hex``

File: source/grid/grid_generator.cc [hunk 4]
```diff
@@ -749,6 +749,63 @@ namespace GridGenerator
...
+            ExcMessage("vertices[6][2] must be greater or equal than vertices[2][2]"));
+    Assert (vertices[7][2]>=vertices[3][2],
+            ExcMessage("vertices[7][2] must be greater or equal than vertices[3][2]"));
+#endif
+
+    // First create an hyper_rectangle and then deform it.
+    Point<dim> p;
+    for (int i=0; i<dim; ++i)
+      p[i] = 1.;
+    hyper_rectangle(tria, Point<dim>(), p, colorize);
```
Comment (tjhei) [inline source/grid/grid_generator.cc]: why not a hyper cube?

File: source/grid/grid_generator.cc [hunk 5]
```diff
@@ -749,6 +749,63 @@ namespace GridGenerator
...
+    Assert (vertices[7][1]>=vertices[5][1],
+            ExcMessage("vertices[7][1] must be greater or equal than vertices[5][1]"));
+    Assert (vertices[4][2]>=vertices[0][2],
+            ExcMessage("vertices[4][2] must be greater or equal than vertices[0][2]"));
+    Assert (vertices[5][2]>=vertices[1][2],
+            ExcMessage("vertices[5][2] must be greater or equal than vertices[1][2]"));
+    Assert (vertices[6][2]>=vertices[2][2],
+            ExcMessage("vertices[6][2] must be greater or equal than vertices[2][2]"));
+    Assert (vertices[7][2]>=vertices[3][2],
+            ExcMessage("vertices[7][2] must be greater or equal than vertices[3][2]"));
```
Comment (tjhei) [inline source/grid/grid_generator.cc]: urgh, this is kind of ugly. :-(

File: tests/grid/grid_generator_hyper_quadrilateral.cc [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+// Test output for GridGenerator::hyper_quadraliteral()
```
Comment (tjhei) [inline tests/grid/grid_generator_hyper_quadrilateral.cc]: quadrilateral :-)

File: include/deal.II/grid/grid_generator.h [hunk 7]
```diff
@@ -292,6 +292,20 @@ namespace GridGenerator
           const std::vector<unsigned int> &holes);
 
   /**
+   * A general quadrilateral in 2d or a general hexahedron in 3d. It is the
+   * responsability of the user to provide the vertices in the right order (see
+   * the documentation of the GeometryInfo class).
+   *
+   * @author Bruno Turcksin
+   * @date 2016
```
Comment (tjhei) [inline include/deal.II/grid/grid_generator.h]: I find "@date" superfluous.

**Comment by davydden** [comment]
i like `general_cell`

**Comment by Rombur** [comment]
Done. It's kinda sad that github does not collapse the comments that have been answered anymore :-(

**Comment by davydden** [comment]
> It's kinda sad that github does not collapse the comments that have been answered anymore :-(

yeap, i realised it in another PR. Very inconvenient...

File: include/deal.II/grid/grid_generator.h [hunk 8]
```diff
@@ -292,6 +292,23 @@ namespace GridGenerator
           const std::vector<unsigned int> &holes);
 
   /**
+   * A general quadrilateral in 2d or a general hexahedron in 3d. It is the
```
Comment (tjhei) [inline include/deal.II/grid/grid_generator.h]: what happens in 1d?

File: include/deal.II/grid/grid_generator.h [hunk 9]
```diff
@@ -292,6 +292,23 @@ namespace GridGenerator
...
+   * responsibility of the user to provide the vertices in the right order (see
+   * the documentation of the GeometryInfo class). If the argument @p
```
Comment (tjhei) [inline include/deal.II/grid/grid_generator.h]: new paragraph for colorize?

File: include/deal.II/grid/grid_generator.h (hunk 8 above)
Comment (Rombur) [inline include/deal.II/grid/grid_generator.h]: You get a line.

File: include/deal.II/grid/grid_generator.h (hunk 9 above)
Comment (Rombur) [inline include/deal.II/grid/grid_generator.h]: I just copied it from `hyper_cube`. You want me to change it there too?

Comment (tjhei) [inline include/deal.II/grid/grid_generator.h]: If you want to and think it makes sense, sure. It was just a suggestion.

**Comment by Rombur** [comment]
OK done.

File: include/deal.II/grid/grid_generator.h (hunk 3 above)
Comment (bangerth) [inline include/deal.II/grid/grid_generator.h]: `cell_from_vertices`

File: source/grid/grid_generator.cc (hunk 5 above)
Comment (bangerth) [inline source/grid/grid_generator.cc]: Why do these assertions make sense? Imagine you wanted to just create a rotated square, then several of these assertions are going to trigger.

You make an assumption here about the alignment of the cell with some kind of coordinate system. I think you should drop this and instead just test that the volume of the cell is positive. I think there are functions in `GridTools` that check that.

File: source/grid/grid_generator.cc [hunk 10]
```diff
@@ -749,6 +749,60 @@ namespace GridGenerator
...
+#endif
+
+    // First create an hyper_rectangle and then deform it.
+    hyper_cube(tria, 0, 1, colorize);
+
+    typename Triangulation<dim>::active_cell_iterator cell = tria.begin_active();
+    for (unsigned int i=0; i<GeometryInfo<dim>::vertices_per_cell; ++i)
+      {
+        Point<dim> &v = cell->vertex(i);
+        v = vertices[i];
```
Comment (bangerth) [inline source/grid/grid_generator.cc]: Just write these two in one line :-)

**Comment by bangerth** [comment]
I like this function. With C++11-style initializers, you will be able to write the whole mesh generation in one statement.

**Comment by Rombur** [comment]
done

File: source/grid/grid_generator.cc [hunk 11]
```diff
@@ -749,6 +749,33 @@ namespace GridGenerator
...
 
+
+
+  template <int dim>
+  void
+  general_cell (Triangulation<dim> &tria,
+                const std::vector<Point<dim> > &vertices,
+                const bool colorize)
+  {
+    Assert (vertices.size() == std::pow(2., dim),
```
Comment (bangerth) [inline source/grid/grid_generator.cc]: Write the rhs as `GeometryInfo<dim>::vertices_per_cell`. That's an integer, and its name already gives away what it is supposed to be.

File: source/grid/grid_generator.cc [hunk 12]
```diff
@@ -749,6 +749,33 @@ namespace GridGenerator
...
+            ExcMessage("Wrong number of vertices."));
+
+    // First create an hyper_rectangle and then deform it.
```
Comment (bangerth) [inline source/grid/grid_generator.cc]: an -> a

File: source/grid/grid_generator.cc [hunk 13]
```diff
@@ -749,6 +749,33 @@ namespace GridGenerator
...
+
+    typename Triangulation<dim>::active_cell_iterator cell = tria.begin_active();
+    for (unsigned int i=0; i<GeometryInfo<dim>::vertices_per_cell; ++i)
+      cell->vertex(i) = vertices[i];
+
+    // Check that the order of the vertices makes sense, i.e., the volume of the
+    // cell is positive.
+    Assert(GridTools::volume(tria) > 0.,
+           ExcMessage(std::string("The volume of the cell is not greater than zero. ") +
+                      "This could be due to the wrong ordering of the vertices."));
```
Comment (bangerth) [inline source/grid/grid_generator.cc]: I think you can just write
```
     ExcMessage ("something "
                           "something else that the C preprocessor automatically collates "
                           "with the first string")
```

**Pull request merged**